### PR TITLE
Ftr/support register with no pojo object

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
**What this PR does**:
Add not pojo object register support, to support dubbo-go-cli tool.
Before, we can't register an object that made by "reflect.New" because it's not impl pojo interface.
But now, we can register this object, as long as it contains field "JavaClassName"

**Which issue(s) this PR fixes**:

NONE
**Special notes for your reviewer**:
Create a function to deal with interface made by reflect, get JavaClassName field and register the object with javaClassName.

**Does this PR introduce a user-facing change?**:
NONE

### CODE REVIEW

**Review by AlexStocks — APPROVED** [review]

**Comment by wongoo** [comment]
@LaurenceLiZhixin  I think the func can meet your requirement ： `func RegisterPOJOMapping(javaClassName string, o interface{}) int `

**Comment by LaurenceLiZhixin** [comment]
> 
> 
> @LaurenceLiZhixin I think the func can meet your requirement ： `func RegisterPOJOMapping(javaClassName string, o interface{}) int `
I think this function can register object, and I actually use this funciton in my code.
But if I want to encode a none POJO obj, hessian2 can't find out what the object is, because it can't identify the object without JavaClassName. That is what my codes do.

**Comment by LaurenceLiZhixin** [comment]
> 
> 
> @LaurenceLiZhixin I think the func can meet your requirement ： `func RegisterPOJOMapping(javaClassName string, o interface{}) int `

I think this function can register object, and I actually use this funciton in my code.
But if I want to encode a none POJO obj, hessian2 can't find out what the object is, because it can't identify the object without JavaClassName. That is what my codes do: find the JavaClassName field and use it to encode registered object.

**Comment by wongoo** [comment]
@LaurenceLiZhixin  
1. it's not efficient to check whether it's exist a field named JavaClassName, u should call checkPOJORegistry first to check whether it's being registered or not.  
2. SHOULD NOT copy a whole method and just change a little logic in it,  a better way is to create a new func `encNamedPOJOObject(javaClassName string, v interface{})` , and the original func `encPOJOObject` calling the new func.
3. Finally, I doubt it's a common way to define JavaClassName in fields, In most situation, it will cause slow performance. So I will not approval this pr.

File: object.go [hunk 1]
```diff
@@ -135,8 +135,10 @@ func (e *Encoder) encObject(v POJO) error {
 		if !ok {
 			if reflect.TypeOf(v).Implements(javaEnumType) {
 				idx = RegisterJavaEnum(v.(POJOEnum))
-			} else {
-				idx = RegisterPOJO(v)
+			} else if isPojo{
+				idx = RegisterPOJO(pojo)
+			}else {
```
Comment (AlexStocks) [inline object.go:140]: gofmt

File: object.go [hunk 2]
```diff
@@ -135,8 +135,10 @@ func (e *Encoder) encObject(v POJO) error {
 		if !ok {
 			if reflect.TypeOf(v).Implements(javaEnumType) {
 				idx = RegisterJavaEnum(v.(POJOEnum))
-			} else {
-				idx = RegisterPOJO(v)
+			} else if isPojo{
```
Comment (AlexStocks) [inline object.go:138]: gofmt

File: object.go [hunk 3]
```diff
@@ -196,104 +198,6 @@ func (e *Encoder) encObject(v POJO) error {
 
 	return nil
 }
-
-func (e *Encoder) encNotPOJOObject(v interface{}, javaClassName string) error {
```
Comment (AlexStocks) [inline object.go]: why delete this func?

Comment (LaurenceLiZhixin) [reply object.go]: As wongoo commented above, I merge the 'no pojo object' register logic in  encPOJOObject function.

**Review by AlexStocks — APPROVED** [review]

File: object.go [hunk 4]
```diff
@@ -124,7 +124,7 @@ func (e *Encoder) encObject(v POJO) error {
 	// write object definition
 	idx = -1
 	for i = range e.classInfoList {
-		if v.JavaClassName() == e.classInfoList[i].javaName {
+		if isPojo && pojo.JavaClassName() == e.classInfoList[i].javaName {
```
Comment (wongoo) [inline object.go:127]: none POJO should also be checked

Comment (LaurenceLiZhixin) [reply object.go]: fixed

File: encode.go [hunk 5]
```diff
@@ -154,6 +154,9 @@ func (e *Encoder) Encode(v interface{}) error {
 				return e.encObject(p)
 			}
 
+			if _, ok := checkPOJORegistry(vv.Type().String()); ok {
```
Comment (wongoo) [inline encode.go:157]: just call `encObject` and not need to check, which will be done in func `encObject`

File: object.go [hunk 6]
```diff
@@ -18,6 +18,7 @@
 package hessian
 
 import (
+	"fmt"
```
Comment (wongoo) [inline object.go:21]: remove it if not being used

Comment (LaurenceLiZhixin) [reply object.go]: fixed

File: object.go [hunk 7]
```diff
@@ -106,8 +106,19 @@ func (e *Encoder) encObject(v POJO) error {
...
-
+	pojo, isPojo := v.(POJO)
 	vv := reflect.ValueOf(v)
+
+	// get none pojo JavaClassName
+	var nonePojoJavaName string
+	if !isPojo {
+		s, ok := pojoRegistry.registry[vv.Type().String()]
+		if !ok {
+			return perrors.New("Not pojo obj:" + typeof(v) + " didn't registered before!")
```
Comment (wongoo) [inline object.go:117]: using `perrors.Errorf` instead

File: object.go [hunk 8]
```diff
@@ -106,8 +106,19 @@ func (e *Encoder) encObject(v POJO) error {
...
 		clsDef classInfo
 	)
-
+	pojo, isPojo := v.(POJO)
 	vv := reflect.ValueOf(v)
+
+	// get none pojo JavaClassName
+	var nonePojoJavaName string
+	if !isPojo {
+		s, ok := pojoRegistry.registry[vv.Type().String()]
```
Comment (wongoo) [inline object.go:115]: there will be a concurrent access issue, I add a func  `loadPOJORegistry` in https://github.com/apache/dubbo-go-hessian2/pull/244/files

File: object.go (hunk 7 above)
Comment (LaurenceLiZhixin) [reply object.go]: fixed

File: object.go (hunk 8 above)
Comment (LaurenceLiZhixin) [reply object.go:115]: I 'll merge master after you merge this pr. And fix again.

Comment (wongoo) [reply object.go:115]: @LaurenceLiZhixin u can fix it now

Comment (LaurenceLiZhixin) [reply object.go]: fixed

**Review by wongoo — APPROVED** [review]
LGTM

**Review by zouyx — APPROVED** [review]

**Pull request merged**